TASK: Introduce api to use global-state for global registry

mhsdesign opened the pull request:
TASK: Introduce api to use global state for configuration and global registry

Originally the Neos Ui was build seemingly without real global state. There is a high order component function - invoked via a decorator `@neos()` which allowed to inject the registry which was distributed via react context.

For functional components currently one would need to use the same high-order-pattern which is a little quirky with functions compared to say the new concept of hooks.

The use of hooks as approach was shown in [`@sitegeist-archeaopteryx/neos-bridge`](https://github.com/sitegeist/Sitegeist.Archaeopteryx/tree/2.0/Neos.Ui/neos-bridge) we would continue to use the `NeosContext` and fetch the global registry or configuration from there via hooks like

- `useNeos()`
- `useGlobalRegistry()`
- `useI18n()`
- `useConfiguration()`


but as this information is truly global and never leads to a re-rendering e.g. is static. We dont have to make that look like a hook but just use javascript state instead. (Like we do for the new `translate()` function)

This change introduces a new package "@neos-project/neos-ui-registry" to host the global registry as singleton

```ts
import {getGlobalRegistry, getRegistryById} from '@neos-project/neos-ui-registry';

const globalRegistry = getGlobalRegistry();
globalRegistry.get('...');
globalRegistry.set('...', ...);

// alias as "get" is common enough
const validators = getRegistryById('validators');
```

### Plugins

Further as part of this refactoring to move correlated code better together - the `SynchronousRegistry` and `SynchronousMetaRegistry` are moved to the new `@neos-project/neos-ui-registry` package.

This has an implication as those are exposed via extensibility api - plugins using `import {SynchronousRegistry} from '@neos-project/neos-ui-extensibility'` must update this to use `import {SynchronousRegistry} from '@neos-project/neos-ui-registry'`. This is a soft breaking change as we do not require anyone to update their `@neos-project/neos-ui-extensibility` package. Previously build plugins continue to work. BUT to also allow plugins build this new way a patch must be made to 9.0 which provides `@neos-project/neos-ui-registry` already for plugins.

-> or we decide to target 9.0 instead.

### Typings 🥳 

Also we now have improved typescript support for the `globalRegistry` - at least with a few registries:

<img width="545" height="202" alt="image" src="https://github.com/user-attachments/assets/27fd2a8b-d03b-4d7a-b51c-a4513b6ce9be" />


**What I did**

**How I did it**

**How to verify it**

<!--
If possible, a screenshot or a gif comparing the new and old behavior would be great.
-->

mhsdesign committed:
TASK: Introduce api to use global state for global registry

Originally the Neos Ui was build seemingly without real global state. There is a high order component function - invoked via a decorator @neos() which allowed to inject the registry.

For functional components currently one would need to use the same high-order-pattern which is a little quirky with functions compared to say the new concept of hooks.

The use of hooks as approach was shown in `@neos-project/neos-ui-link-editor-neos-bridge` we would continue to use the NeosContext and fetch the global registry or configuration from there via hooks like

- useNeos()
- useGlobalRegistry()
- ...useConfiguration()

but as this information is truly global and never leads to a re-rendering e.g. is static. We dont have to make that look like a hook but just use javascript state instead. (Like we do for the new translate() function)

mhsdesign committed:
TASK: Introduce dedicated package `@neos-project/neos-ui-registry` to encapsulate global registry

mhsdesign committed:
TASK: Move `SynchronousRegistry` and `SynchronousMetaRegistry` to new `@neos-project/neos-ui-registry` package

and alias and expose it via extensibility api - plugins using `import {SynchronousRegistry} from '@neos-project/neos-ui-extensibility'` must update this to use `import {SynchronousRegistry} from '@neos-project/neos-ui-registry'`

mhsdesign committed:
TASK: Add correct typescript typing to `manifest()` function

mhsdesign committed:
TASK: Move `GlobalRegistry` type to `neos-ui-registry` and use ts module overloading for extending it to know "i18n" etc

mhsdesign committed:
TASK: Move `GlobalRegistry` "i18n" overload to i18n package

mhsdesign committed:
TASK: Move `GlobalRegistry` "validators" overload to validators package

... and also instantiate validators registry here

mhsdesign committed:
TASK: Also re-export `SynchronousMetaRegistry` as its used in the neos-ui package

mhsdesign committed:
TASK: Solve `SynchronousMetaRegistry` vs `GlobalRegistry` type riddle

mhsdesign committed:
TASK: Move `NodeTypesRegistry` type to `@neos-project/neos-ui-contentrepository` package and register type in global registry

mhsdesign committed:
PATCH: Fix typescript type for `referenceGrandParentNodeType` to allow null

as called with

```
const referenceGrandParentNodeType = referenceParentNode ? referenceParentNode.nodeType : null;
```

mhsdesign committed:
PATCH: Fix dangerous ! double quotes

Hi style ci do you see this:

"""""""""""""""""""""""""""""""""""""""""""""""""""""

hope youre "happy"

mhsdesign committed:
PATCH: Fix NMP build of `@neos-project/neos-ui-extensibility` by separating types declaration

otherwise typescript attempts to parse the whole neos ui repo and that fails

mhsdesign commented:
In some (minor) regards this change itself is not finished -> it definitely needs the followup https://github.com/neos/neos-ui/pull/4012 but id prefer to merge these as separate parts

Sebobo commented on packages/neos-ui-extensibility/src/index.ts:
```diff
@@ -1,15 +1,30 @@
 export {
     createConsumerApi,
-    readFromConsumerApi,
-    SynchronousRegistry,
```
Removing those exports would be breaking in the extensibility package. I have a plugin that imports them.

mhsdesign replied:
im curious how does one use `readFromConsumerApi`? 😅 that is accidental imo?

Sebobo replied:
I meant the registry

Sebobo reviewed (changes requested):
Looks good, but we have to be careful with any changes to the extensibility package

mhsdesign commented:
Regarding that we no longer expose `SynchronousRegistry` when doing `import {SynchronousRegistry} from '@neos-project/neos-ui-extensibility'` -> i know that this might cause adjustments in the plugins and i also wrote about that above under "Plugins".

At first i thought we could provide a b/c layer to still support that but that might be impossible due to the global state initialisation.

I think the important part is that few people used `SynchronousRegistry` and that already build plugins continue to work and that if people update their extensibility yarn package `@neos-project/neos-ui-extensibility` from 9.0.2 to this planned upcoming 9.0.3 that yes they have to rewrite the import to `import {SynchronousRegistry} from '@neos-project/neos-ui-registry'` -> but i believe this might be preferred over to targeting that change only to 9.1 which would mean that plugins build with `'@neos-project/neos-ui-registry'`` would not work on a Neos 9.0 host ... if we merge into 9.0.3 these plugins would at least work with Neos 9.0.3 and above.

If we consider this a bad deal still and want to preserve full yarn semver promise i will find a way to make `import {SynchronousRegistry} from '@neos-project/neos-ui-extensibility'` work -> even if we need to duplicate code.

EDIT @Sebobo found a way to support the "SynchronousRegistry" still  744659a93fd27dba763308967a0b55e0cf9e10ab

mhsdesign committed:
TASK: Dont expose ui-internal createConsumerApi and readFromConsumerApi for plugins but preserver compatibility with `SynchronousRegistry` legacy import

Sebobo commented:
> EDIT @Sebobo found a way to support the "SynchronousRegistry" still [744659a](https://github.com/neos/neos-ui/commit/744659a93fd27dba763308967a0b55e0cf9e10ab)

Thx. Cool, less breaky, more happy and yes I agree, we should keep targeting 9.0 with this.

mhsdesign reviewed (commented):
I applied some git magic to backport some of the fixups from the other pr to here ;) Now this pr is independent ;)

And as you agree ill probably merge it soon before the sprint^^

mhsdesign committed:
PATCH: Remove dependency to `@neos-project/positional-array-sorter` in "neos-ui-extensibility"

this is because the SynchronousRegistry is no longer shipped as plain code using the "positional-array-sorter" but exposed via the consumer api

mhsdesign committed:
TASK: Document plugin-api.js
